pkg/k8shandler/recovery.go

```diff
@@ -0,0 +1,315 @@
+			// keep track of the index it was at so we can correctly update it later
+			nodesToMatch[nodeIndex] = node
+		}
+	}
+
+	if len(nodesToMatch) > 0 {
+		// first get a list of known UUIDs
+		knownUUIDs := []string{}
+		for _, node := range er.cluster.Spec.Nodes {
+			if node.GenUUID != nil {
+				knownUUIDs = append(knownUUIDs, *node.GenUUID)
+			}
+		}
+
+		// collect uuid counts
+		selector := map[string]string{}
+		pvcList, err := GetPVCList(er.cluster.Namespace, selector, er.client)
+		if err != nil {
+			// TODO: handle it
+		}
+
+		if len(pvcList.Items) > 0 {
+			er.recoverFromPVCs(knownUUIDs, nodesToMatch)
+		} else {
+			er.recoverFromDeployments(knownUUIDs, nodesToMatch)
+		}
+	}
+}
+
+func (er *ElasticsearchRequest) deploymentSpecMatchNode(node api.ElasticsearchNode, deployment apps.DeploymentSpec, count int32) bool {
+
+	// we know roles will match, first check node count
+	if node.NodeCount != count {
+		return false
+	}
+
+	return er.podSpecMatchNode(node, deployment.Template.Spec)
+}
+
+func (er *ElasticsearchRequest) statefulSetSpecMatchNode(node api.ElasticsearchNode, statefulset apps.StatefulSetSpec) bool {
+
+	// we know roles will match, first check node count
+	if node.NodeCount != *statefulset.Replicas {
+		return false
+	}
+
+	return er.podSpecMatchNode(node, statefulset.Template.Spec)
+}
+
+func (er *ElasticsearchRequest) podSpecMatchNode(node api.ElasticsearchNode, podSpec v1.PodSpec) bool {
+
+	selectors := mergeSelectors(node.NodeSelector, er.cluster.Spec.Spec.NodeSelector)
+	selectors = utils.EnsureLinuxNodeSelector(selectors)
+
+	if !areSelectorsSame(selectors, podSpec.NodeSelector) {
+		return false
+	}
+
+	tolerations := appendTolerations(node.Tolerations, er.cluster.Spec.Spec.Tolerations)
+
+	if !containsSameTolerations(podSpec.Tolerations, tolerations) {
+		return false
+	}
+
+	nodeResources := newESResourceRequirements(node.Resources, er.cluster.Spec.Spec.Resources)
+	proxyResources := newESProxyResourceRequirements(node.ProxyResources, er.cluster.Spec.Spec.ProxyResources)
+
+	var deploymentNodeResources v1.ResourceRequirements
+	var deploymentProxyResources v1.ResourceRequirements
+
+	for _, container := range podSpec.Containers {
+		if container.Name == "elasticsearch" {
+			deploymentNodeResources = container.Resources
+		} else {
+			deploymentProxyResources = container.Resources
+		}
+	}
+
+	// make sure resources are same
+	if different, _ := utils.CompareResources(nodeResources, deploymentNodeResources); different {
+		return false
+	}
+
+	// make sure proxy resources are same
+	if different, _ := utils.CompareResources(proxyResources, deploymentProxyResources); different {
+		return false
+	}
+
+	return true
+}
+
+func parseNodeName(name string) (clusterName, roles, uuid string, replica int) {
+
+	splitName := strings.Split(name, "-")
+	if len(splitName) != 4 {
+		return
+	}
+
+	clusterName = splitName[0]
+	roles = splitName[1]
+	uuid = splitName[2]
+	replica, err := strconv.Atoi(splitName[3])
```

Comment:
AFAIK, this naming schema is our convention, but since we don't declare `ownerReferences` for PVCs, we may hit a cluster having arbitrary PVCs with this schema but alien data for Elasticsearch. I believe this is a super rare case, but wouldn't be an issue if we use `ownerReferences`. WDYT?

Reply:
so the reason we never added `OwnerReferences` for PVC volumes was to prevent having the scheduler remove them. In fact, if we had added them we would be in a different situation currently -- customer's PVCs would have been deleted along with their old cluster resulting in a total loss of data. I did consider adding labels for our current PVCs, however that doesn't help with customer currently facing this issue (unless we ask them to label the PVCs they want to use...) maybe that will be the route to take? label the ones they want to keep and delete the rest...

Reply:
i can also add some sanity checking in here to make sure the pvc name matches the pattern we use.

Reply:
:facepalm: IC ownerReferences are always a double-edged sword, except if we instruct a special finalizer not deleting them. This would too much of a hazzle.

Labels should add some extra sanity but they have the same issues as the naming schema. Nobody can stop users using the same labels for non-ES PVCs.

Adding some extra sanity on name checking would be fine for me.

Reply:
we'll have both the labels to do a `get` on and the cluster name verification :)